### ISSUE TITLE
`%callback` unit test seems wrong

### ISSUE THREAD

**Issue by mmomtchev** [swig/swig#2590]
Currently, JavaScript does not run the `%callback` unit test - as it is enabled on a per-language basis. Yet, there is a `callback_runme.js` to be found in the JavaScript test-suite:

https://github.com/swig/swig/blob/f574a34155e4ede49b1b023494e0a1637ab89f8f/Examples/test-suite/javascript/callback_runme.js

As far as I can see, from the existing documentation, this is not expected to work:

https://github.com/swig/swig/blob/ef112087cc22d4d7d6041d159834c27d56eb7984/Examples/test-suite/callback.i

This interface file tries to make a global entity `foo` that can be used both as a function - when called from JS - and as a constant - when passed to C/C++ as a pointer. This might be possible with NAPI - but it will require extending the `swig_type_info` and it seems to be a really complex feature without any real benefit.

Of all the languages, only Python, JavaScript, PHP and Octave have implemented this test. JavaScript and Octave do not actually run it. Python has a special Python-specific implementation. PHP implements only the last part of the test - which is possible in JavaScript too.

Is this test still valid or has the feature been more or less abandoned in favor of directors? Many backends produce code that does not compile or even crash on this test.

**Comment by ojwb** [comment]
Thanks for raising this.

While `%callback` serves a vaguely similar purpose to directors, they aren't really usable as replacements for each other.  `%callback` is useful for C APIs with API functions which take a callback function as a parameter and also supply suitable functions to pass (notably they don't support passing a target language function for the callback AFAICS), while directors are useful for C++ APIs where there are API classes designed to be subclassed and support implementing such a subclass in the target language.

They seem to still be maintained, e.g. 3aa302c08f609ad15a602b32a8c0124eb48861f9 fixed a `%callback` issue just last year.

> PHP implements only the last part of the test - which is possible in JavaScript too.

Looks like that was originally in the `autodoc` testcase but got relocated to a more appropriate place: a4b319ce8eba4fe971235f611fd656148f6f3ed5

Looking at `callback_runme.py`, it seems to assume one can call the callback directly as well as passing it but that doesn't match the docs for `%callback`:

```
 If you want to make a function available as both a callback function and a function, you can use the %callback and %nocallback directives like this:

/* Function with a callback */
int binary_op(int a, int b, int (*op)(int, int));

/* Some callback functions */
%callback("%s_cb");
int add(int, int);
int sub(int, int);
int mul(int, int);
%nocallback;

The argument to %callback is a printf-style format string that specifies the naming convention for the callback constants (%s gets replaced by the function name). The callback mode remains in effect until it is explicitly disabled using %nocallback. When you do this, the interface now works as follows:

>>> binary_op(3, 4, add_cb)
7
>>> binary_op(3, 4, mul_cb)
12
>>> add(3, 4)
7
>>> mul(3, 4)
12

Notice that when the function is used as a callback, special names such as add_cb are used instead. To call the function normally, just use the original function name such as add(). 
```

There used to be a separate `%pythoncallback` but they effectively got merged years ago it seems, with the remaining quirk that SWIG/Python supports `%callback(1)` but other target languages don't.  I think that reflects that with Python it's possible to have a single value be both the callback constant and function, whereas in other target languages it isn't (or at least the `%callback` implementations aren't written to allow it).

So it seems `callback.i` really ought to use `%nocallback` and it should rename the callback functions as the docs say, at least for the ones it expects to work as both.  I'll see if I can fix that.

I think it's reasonable to regard `%callback` as a non-core feature which it's OK for target languages not to support, though it seems the problem for JS (and indeed probably most target languages) is the testcase trying to use the same name.

I've come across a few `*_runme.*` files which were actually unused over the years, some of which worked fine (or with only a minor update) when enabled so presumably were just failures to actually add the testcase to the list to run while others of which seemed to be effectively a failed attempt to implement.

**Comment by ojwb** [comment]
I tried updating `callback.i` to rename, but with this patch applied

```
diff --git a/Examples/test-suite/callback.i b/Examples/test-suite/callback.i
index e406615b..10887609 100644
--- a/Examples/test-suite/callback.i
+++ b/Examples/test-suite/callback.i
@@ -10,7 +10,7 @@
 %callback("%s") foo;
 %callback("%s") foof;
 %callback("%s") A::bar;
-%callback("%s") A::foom;
+%callback("%s_cb") A::foom;
 #endif
 %callback("%(uppercase)s_Cb_Ptr") foo_T;  // this works in Python too
 %callback("%s_cb") identity_finger;
```

running the callback testcase for PHP makes SWIG fail an assertion and crash:

```
$ make callback.cpptest
if [ -f ./callback_runme.php ]; then echo "checking php testcase callback (with run test)" ; else echo "checking php testcase callback" ; fi
checking php testcase callback (with run test)
make -f ../../../Examples/Makefile SRCDIR='./' CXXSRCS='' SWIG_LIB_DIR='../../../Lib' SWIGEXE='../../../swig' INCLUDES='-I../../../Examples/test-suite' SWIGOPT='-outcurrentdir -I../../../Examples/test-suite' NOLINK=true TARGET='callback' INTERFACEDIR='../' INTERFACE='callback.i' php_cpp
make[1]: Entering directory '/home/olly/git/swig/Examples/test-suite/php'
env SWIG_LIB=../../../Lib  ../../../swig -php -c++ -outcurrentdir -I../../../Examples/test-suite -o callback_wrap.cxx ./../callback.i
swig: Modules/emit.cxx:407: int emit_action_code(Node*, String*, String*): Assertion `Getattr(n, "wrap:name")' failed.
make[1]: *** [../../../Examples/Makefile:1224: php_cpp] Aborted (core dumped)
make[1]: Leaving directory '/home/olly/git/swig/Examples/test-suite/php'
make: *** [Makefile:52: callback.cpptest] Error 2
```

I don't really have time or energy to go further down this rabbit hole right now, but I still think this testcase should be adjusted to be portable to target languages aside from Python.

**Comment by Andy-zhujunwen** [comment]
are swig cannot produce the director of javascript?

**Comment by ojwb** [comment]
> are swig cannot produce the director of javascript?

No, but that's not what this ticket is about and there's a separate ticket for it: #1244
